File: Conjugate_gradient/cg.r
conjugate_gradient <- function(A, b, tol = 1e-6, max_iter = 1000) {
  n <- length(b)
  x <- numeric(n)  # Initial guess (0,0,0)
  r <- b - A %*% x  # Initial residual
  p <- r  # Initial search direction
  rsold <- crossprod(r)  # r^T r (crossprod is more efficient)
  
  for (iter in 1:max_iter) {
    Ap <- A %*% p  # Matrix-vector product
    alpha <- as.numeric(rsold / crossprod(p, Ap))  # Step size
    x <- x + alpha * p  # Update solution
    r <- r - alpha * Ap  # Update residual
    rsnew <- crossprod(r)  # New residual norm
    
    # Check convergence
    if (sqrt(rsnew) < tol) {
      cat(sprintf("Converged in %d iterations\n", iter))
      break
    }
    
    # Update search direction
    beta <- as.numeric(rsnew / rsold)
    p <- r + beta * p
    rsold <- rsnew
  }
  
  if (sqrt(rsnew) >= tol) {
    cat(sprintf("Did not converge in %d iterations\n", max_iter))
  }
  
  return(x)
}

# Define matrix A and vector b
A <- matrix(c(4, -1, 0,
              -1, 4, -1,
              0, -1, 4), nrow = 3, byrow = TRUE)
b <- c(1, 2, 3)

# Solve using Conjugate Gradient
x <- conjugate_gradient(A, b)

# Display solution
cat("\nCG Solution:\n")
print(x)

# Validate solution
residual <- A %*% x - b
residual_norm <- norm(residual, "2")
cat(sprintf("\nResidual norm: %.2e\n", residual_norm))

# Compare with exact solution
x_exact <- solve(A, b)
cat("Exact solution (solve(A, b)):\n")
print(x_exact)

# Calculate error norm
error_norm <- norm(x - x_exact, "2")
cat(sprintf("\nError norm: %.2e\n", error_norm))

File: Conjugate_gradient/cg.jl
using LinearAlgebra

# Define the tridiagonal matrix and vector
A = Tridiagonal([-1.0, -1.0], [4.0, 4.0, 4.0], [-1.0, -1.0])
b = [1.0, 2.0, 3.0]

function conjugate_gradient(A, b; tol=1e-12, max_iter=1000)
    x = zeros(length(b))    # Initial guess
    r = copy(b)             # Initial residual (b - A*0)
    p = copy(r)             # Initial search direction
    rsold = dot(r, r)       # Initial residual squared norm
    
    for iter in 1:max_iter
        Ap = A * p          # Matrix-vector product
        alpha = rsold / dot(p, Ap)
        x .+= alpha .* p    # Update solution
        r .-= alpha .* Ap   # Update residual
        rsnew = dot(r, r)   # New residual norm
        
        # Check convergence
        if sqrt(rsnew) < tol
            println("Converged in $iter iterations")
            break
        end
        
        # Update search direction
        beta = rsnew / rsold
        p .= r .+ beta .* p
        rsold = rsnew
    end
    return x
end

# Solve using Conjugate Gradient
x = conjugate_gradient(A, b)

# Display results
println("\nCG Solution:")
println(x)

# Validate solution
residual = A * x - b
println("\nResidual norm: ", norm(residual))

# Compare with Julia's built-in solver
x_exact = A \ b
println("\nExact solution:")
println(x_exact)

# Calculate error norm
println("\nError norm: ", norm(x - x_exact))
